client/src/main/java/io/pravega/client/stream/impl/EventStreamWriterImpl.java

```diff
@@ -122,10 +153,21 @@ private void handleMissingLog() {
      * message find which new segment it should go to and send it there. This can happen recursively if segments turn
      * over very quickly.
      */
-    @GuardedBy("lock")
     private void handleLogSealed(Segment segment) {
-        List<PendingEvent> toResend = selector.refreshSegmentEventWritersUponSealed(segment);
-        resend(toResend);
+        lock.readLock().lock();
+        /* Using segmentSealedLock the following behaviour is enforced
+            - Prevent concurrent segmentSealedCallback for different segments from being invoked concurrently.
+            - Ensure waiting segmentSealedCallbacks are invoked before the next write is invoked.
+            This ensures that resend() would be invoked again if we observe a segment sealed exception.
+         */
+        segmentSealedLock.lock();
```

Comment:
The easiest way to spell that out, and detect that via static analysis tools would be to annotate SegmentSealedLock with @GuardedBy("lock")